src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -759,5 +763,90 @@ private static bool IsPotentialFilterCharacter(char c)
+            if (defaults.IsDefaultOrEmpty || itemsWithMatch[selectedIndex].item.Rules.MatchPriority >= MatchPriority.Preselect)
+                return (selectedIndex, false);
+
+            if (aggressive)
+            {
+                foreach (var defaultText in defaults)
+                {
+                    for (var itemIndex = 0; (itemIndex < itemsWithMatch.Length); ++itemIndex)
+                    {
+                        var (item, patternMatch) = itemsWithMatch[itemIndex];
+                        if (item.FilterText == defaultText)
+                        {
+                            if (patternMatch == null || patternMatch.Value.Kind <= PatternMatchKind.Prefix)
+                                return (itemIndex, true);
+
+                            break;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                int similarItemsStart;
+                int dissimilarItemIndex;
+
+                if (filterText.Length == 0)
+                {
+                    // If there is no applicableToSpan, then all items are equally similar.
+                    similarItemsStart = 0;
+                    dissimilarItemIndex = itemsWithMatch.Length;
+                }
+                else
+                {
+                    // Assume that the selectedIndex is in the middle of a range of -- as far as the pattern matcher is concerned --
+                    // equivalent items. Find the first & last items in the range and use that to limit the items searched for from
+                    // the defaults list.                    
+                    var (_, selectedItemMatch) = itemsWithMatch[selectedIndex];
+                    if (!selectedItemMatch.HasValue)
+                        return (selectedIndex, false);
+
+                    similarItemsStart = selectedIndex;
+                    while (--similarItemsStart >= 0)
+                    {
+                        var (_, itemMatch) = itemsWithMatch[similarItemsStart];
+                        if ((!itemMatch.HasValue) || itemMatch.Value.CompareTo(selectedItemMatch.Value) > 0)
+                            break;
+                    }
+
+                    similarItemsStart++;
+
+                    dissimilarItemIndex = selectedIndex;
+                    while (++dissimilarItemIndex < itemsWithMatch.Length)
+                    {
+                        var (_, itemMatch) = itemsWithMatch[dissimilarItemIndex];
+                        if ((!itemMatch.HasValue) || itemMatch.Value.CompareTo(selectedItemMatch.Value) > 0)
+                            break;
+                    }
+                }
+
+                if (dissimilarItemIndex > selectedIndex + 1)
+                {
+                    foreach (var defaultText in defaults)
+                    {
+                        for (var itemIndex = similarItemsStart; (itemIndex < dissimilarItemIndex); ++itemIndex)
+                        {
```

Comment:
how many items may be in 'defaults' and in these loops?  i'm worried about very expensive overhead.

Reply:
@dpugh mentioned there could potentially be multiple, but it's not clear to me how would that work if the main use case for defaults is to keep completion list selection in sync with whole-line completion.

I'm wondering if we can hardcode to only use the first one until we have more clarity on this.